sw/device/lib/base/hardened.h

```diff
@@ -477,36 +477,75 @@ inline uintptr_t ct_cmovw(ct_boolw_t c, uintptr_t a, uintptr_t b) {
   return (launderw(c) & a) | (launderw(~c) & b);
 }
 
+// Implementation details shared across shutdown macros.
+#ifndef OT_OFF_TARGET_TEST
+// This string can be tuned to be longer or shorter as desired, for
+// fault-hardening purposes.
+#define HARDENED_UNIMP_SEQUENCE_() "unimp; unimp; unimp; unimp;"
+
+#define SHUTDOWN_CHECK_OP_eq "beq"
+#define SHUTDOWN_CHECK_OP_ne "bne"
+#define SHUTDOWN_CHECK_OP_ge "bge"
+#define SHUTDOWN_CHECK_OP_le "ble"
+#define SHUTDOWN_CHECK_OP_lt "blt"
+#define SHUTDOWN_CHECK_OP_gt "bgt"
+#define SHUTDOWN_CHECK_(op_, a_, b_) \
+  asm volatile(                         \
+      op_ " %0, %1, .L_SHUTDOWN_EQ_%=;" \
+      HARDENED_UNIMP_SEQUENCE_()        \
+      ".L_SHUTDOWN_EQ_%=:;"             \
+      :: "r"(a_), "r"(b_)               \
+  )
+#define HARDENED_UNREACHABLE_()               \
+  do {                                        \
+    asm volatile(HARDENED_UNIMP_SEQUENCE_()); \
+    __builtin_unreachable();                  \
+  } while (false)
+#else
+#include <assert.h>
+#define SHUTDOWN_CHECK_OP_eq ==
+#define SHUTDOWN_CHECK_OP_ne !=
+#define SHUTDOWN_CHECK_OP_ge >=
+#define SHUTDOWN_CHECK_OP_le <=
+#define SHUTDOWN_CHECK_OP_lt <
+#define SHUTDOWN_CHECK_OP_gt >
+#define SHUTDOWN_CHECK_(op_, a_, b_) assert((a_)op_(b_))
+#define HARDENED_UNREACHABLE_() assert(false)
+#endif
+
 /**
- * Evaluate a manifestly true expression and generate an illegal instruction
- * exception if the result is false.
+ * Indicates that the following code is unreachable; it should never be
+ * reached at runtime.
+ *
+ * If it is reached anyways, an illegal instruction will be executed.
+ */
+#define HARDENED_UNREACHABLE() HARDENED_UNREACHABLE_()
+
+/**
+ * Compare two values in a way that is *manifestly* true: that is, under normal
+ * program execution, the comparison is a tautology.
+ *
+ * If the comparison fails, we can deduce the device is under attack, so an
+ * illegal instruction will be executed. The manner in which the comparison is
+ * done is carefully constructed in assembly to minimize the possibility of
+ * adversarial faults.
+ *
+ * The comparison should be specified as one of the tokens eq, ne, gt, lt, ge,
+ * or le. These correspond to equals, not equals, greater than, less than,
+ * greater than or equal, and less than or equal.
  *
  * This macro is intended to be used along with `launder32()` to handle detected
  * faults when implementing redundant checks, i.e. in places where the compiler
  * could otherwise prove statically unreachable. For example:
  * ```
  * if (launder32(x) == 0) {
- *   SHUTDOWN_CHECK(launder32(x) == 0);
+ *   SHUTDOWN_CHECK(eq, launder32(x), 0);
```

Comment:
nit: How about `SHUTDOWN_CHECK_{EQ,NE...}` sort of matching the approach used in test macros? I feel that it may be a bit more ergonomic.

Reply:
I also think it would be a bit more ergonomic, but I also understand that having `op_` saves us from essentially duplicating the same macro for each type of comparison. @mcy WDYT about defining `CHECK_{EQ, NE, ...}(a_, b_)` as `CHECK_(CHECK_OP_{EQ, NE, ...}, a_, b_)` to have both?

Please also see my comment regarding renaming these to `HARDENED_CHECK_{EQ, NE, ...}` for consistency.

Reply:
I would be ok having six macros. Let's do that.